contract_comm/evm.go

```diff
@@ -18,139 +18,92 @@ package contract_comm
 
 import (
 	"math/big"
-	"reflect"
-	"time"
 
 	"github.com/celo-org/celo-blockchain/accounts/abi"
 	"github.com/celo-org/celo-blockchain/common"
-	"github.com/celo-org/celo-blockchain/common/hexutil"
 	"github.com/celo-org/celo-blockchain/contract_comm/errors"
 	"github.com/celo-org/celo-blockchain/contracts"
 	"github.com/celo-org/celo-blockchain/core/types"
 	"github.com/celo-org/celo-blockchain/core/vm"
-	"github.com/celo-org/celo-blockchain/core/vm/vmcontext"
 	"github.com/celo-org/celo-blockchain/log"
-	"github.com/celo-org/celo-blockchain/metrics"
 )
 
 var (
-	internalEvmHandlerSingleton *InternalEVMHandler
+	systemEVMFactory vm.SystemEVMFactory
 )
 
-// An EVM handler to make calls to smart contracts from within geth
-type InternalEVMHandler struct {
-	chain vm.ChainContext
-}
-
-func MakeStaticCall(registryId [32]byte, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, header *types.Header, state vm.StateDB) (uint64, error) {
-	return makeCallWithContractId(registryId, abi, funcName, args, returnObj, gas, nil, header, state, true)
-}
-
-func MakeCall(registryId [32]byte, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int, header *types.Header, state vm.StateDB, finaliseState bool) (uint64, error) {
-	gasLeft, err := makeCallWithContractId(registryId, abi, funcName, args, returnObj, gas, value, header, state, false)
-	if err == nil && finaliseState {
-		state.Finalise(true)
+func SetSystemEVMFactory(_systemEVMFactory vm.SystemEVMFactory) {
+	if systemEVMFactory == nil {
+		log.Trace("Setting the systemEVMFactory Singleton")
+		systemEVMFactory = _systemEVMFactory
 	}
-	return gasLeft, err
 }
 
-func MakeStaticCallWithAddress(scAddress common.Address, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, header *types.Header, state vm.StateDB) (uint64, error) {
-	return makeCallFromSystem(scAddress, abi, funcName, args, returnObj, gas, nil, header, state, true)
-}
-
-func GetRegisteredAddress(registryId [32]byte, header *types.Header, state vm.StateDB) (common.Address, error) {
-	vmevm, err := createEVM(header, state)
+func MustGetCaller(header *types.Header, state vm.StateDB) vm.SystemEVM {
+	caller, err := getCaller(header, state)
 	if err != nil {
-		return common.ZeroAddress, err
+		panic("failed to get caller")
```

Comment:
If we're going to panic we should at least include the error in the message, or log it before panicking.